README.MD
# 🦸🏻‍♀️ Boca Code Intranet API Project 🦸🏽‍♂️

A place for new students to get to know the area, engage in technology and practice the coding skills that will get you one step closer to becoming a Software Engineer 👩🏼‍💻 🧑🏽‍💻 👩🏾‍💻 🧑🏻‍💻

### From Alumni to Coming Students

This project is built with React, primarily in TypeScript. However, JavaScript contributions are also welcomed.

## Getting Started

1. Copy and paste this command in the command line to download the repo

```bash
$ git clone https://github.com/bocacode/bocacode-intranet-api-typescript.git
```

2. Go to bocacode-intranet-typescript directory

```bash
$ cd bocacode-intranet-api-typescript
```

3. Install third-party packages and libraries

```bash
$ npm install
```

4. Create a .env file and define variables

```bash
MONGO_URI='mongodb://localhost:27020/intranet-data'
PORT=4040
NODE_ENV='development'
```

5. Run Docker client in your machine to build the local MongoDB
6. Build the docker container

```bash
$ docker compose up
```

![docker](https://img.shields.io/badge/Docker-blue?style=for-the-badge&logo=Docker&logoColor=white)

7. Seed data into local MongoDb

```bash
$ npm run seed
```

8. Run project locally

```bash
$ npm run start
```

---
## Unit Tests
----
 
```bash
# to run all unit tests
npm run test 

# to run only unit test for one file
npm run test:watch name_of_file_here
```


# Sample Routes

> Use these samples to test your endpoints and or requests:

## Get all:

```javascript
GET / api / restaurants / HTTP / 1.1
Host: localhost: 4040

// sample response
[
  {
    "active": false,
    "_id": "640a1e8fd56d69fff7b9e3ca",
    "uid": "2335943",
    "address": "7032 W Palmetto Park Rd, Boca Raton, FL 33433",
    "menu_url": "starbucks.com/menu",
    "name": "Starbucks ",
    "photo_url": "https://firebasestorage.googleapis.com/v0/b/bocacode-portal.appspot.com/o/restaurants%2Fstarbucks%20.jpeg?alt=media&token=404140bf-3139-4ad4-bd24-24de58e34ddf",
    "rating": ["4"],
    "created_by": "data seed",
    "__v": 0,
    "createdAt": "2023-03-09T17:59:43.845Z",
    "updatedAt": "2023-03-09T17:59:43.845Z"
  },
  {
    "active": false,
    "_id": "640a1e8fd56d69fff7b9e3cb",
    "uid": "2335943",
    "address": "7028 W Palmetto Park Rd Ste 106, Boca Raton, FL 33433",
    "menu_url": "https://www.chipotle.com/order/#menu",
    "name": "Chipotle ",
    "photo_url": "https://firebasestorage.googleapis.com/v0/b/bocacode-portal.appspot.com/o/restaurants%2Fchipotle.jpeg?alt=media&token=6e0caaf6-82fd-45c4-9ea2-60441d8ff001",
    "rating": ["3"],
    "created_by": "data seed",
    "__v": 0,
    "createdAt": "2023-03-09T17:59:43.845Z",
    "updatedAt": "2023-03-09T17:59:43.845Z"
  }
]
```

## Add One:

```javascript
POST /api/restaurants/ HTTP/1.1
Host: localhost:4040
Content-Type: application/json
Content-Length: 388

{
    "address": "8979 W Palmetto Park Rd, Boca Raton, FL 33433",
    "menu_url": "https://www.chipotle.com/order/#menu",
    "name": "Chipotle 4",
    "photo_url": "https://firebasestorage.googleapis.com/v0/b/bocacode-portal.appspot.com/o/restaurants%2Fchipotle.jpeg?alt=media&token=6e0caaf6-82fd-45c4-9ea2-60441d8ff001",
    "rating": [
        "3"
    ],
    "user_id": "local admin"
}

// sample response
{
    "uid": "236954",
    "address": "8979 W Palmetto Park Rd, Boca Raton, FL 33433",
    "menu_url": "https://www.chipotle.com/order/#menu",
    "name": "Chipotle 4",
    "photo_url": "https://firebasestorage.googleapis.com/v0/b/bocacode-portal.appspot.com/o/restaurants%2Fchipotle.jpeg?alt=media&token=6e0caaf6-82fd-45c4-9ea2-60441d8ff001",
    "rating": [
        "3"
    ],
    "created_by": "admin@gmail.com",
    "active": false,
    "_id": "649715527c16307d70b48737",
    "createdAt": "2023-06-24T16:09:54.535Z",
    "updatedAt": "2023-06-24T16:09:54.535Z",
    "__v": 0
}
```

## Edit One:

```javascript
PATCH /api/restaurants/2335943 HTTP/1.1
Host: localhost:4040
Content-Type: application/json
Content-Length: 95

{
    "address": "8979 W New Address here Boca Raton, FL 33433",
    "user_id": "local admin"
}

// sample response

{
  "active": false,
  "_id": "640a1e8fd56d69fff7b9e3ca",
  "uid": "2335943",
  "address": "8979 W New Address here Boca Raton, FL 33433",
  "menu_url": "starbucks.com/menu",
  "name": "Starbucks ",
  "photo_url": "https://firebasestorage.googleapis.com/v0/b/bocacode-portal.appspot.com/o/restaurants%2Fstarbucks%20.jpeg?alt=media&token=404140bf-3139-4ad4-bd24-24de58e34ddf",
  "rating": ["4"],
  "created_by": "data seed",
  "__v": 0,
  "createdAt": "2023-03-09T17:59:43.845Z",
  "updatedAt": "2023-06-24T15:52:34.555Z"
}
```

## Disable One:

```javascript
PATCH /api/restaurants/disable/236954 HTTP/1.1
Host: localhost:4040
Content-Type: application/json
Content-Length: 36

{
    "user_id": "admin@gmail.com"
}

// sample response

{
    "success": "Restaurant id 236954 has been disabled "
}
```

## Steps:

1. Sign up your user account, by sending a POST request with `RAW` JSON like this

```JSON
// POST to - http://localhost:4040/api/users/signup
{
    "email": "jso1919@gmail.com",
    "password": "mypassword-here",
    "first_name": "Jon",
    "last_name": "Sanchez",
    "access_level": 10
}
```

2. Then user will be added to the Database, with your password hashed, and response should look like this

```JSON
{
    "email": "jso1919@gmail.com",
    "access_level": 10,
    "first_name": "Jon",
    "last_name": "Sanchez",
    "password": "$2b$10$wx2HeoZPXWk7m1JnERlZ/u5PgpRuNANCse5Mzf3MjN5y7DVfrMKCO",
    "status": "Pending",
    "_id": "640a5a268c31ba12e4eab8ef",
    "createdAt": "2023-03-09T22:13:58.848Z",
    "updatedAt": "2023-03-09T22:13:58.848Z",
    "__v": 0,
    "iat": 1678400038,
    "accessToken": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJlbWFpbCI6ImpzbzE5MTlAZ21haWwuY29tIiwiYWNjZXNzX2xldmVsIjoxMCwiZmlyc3RfbmFtZSI6IkpvbmF0aGFuIiwibGFzdF9uYW1lIjoiU2FuY2hleiIsInBhc3N3b3.......................0dNjX6X_yCjdt1PB4Hf4NE1wuvBWCXD5aD9GD72QuDk"
}
```

3. Then you will need to add the `accessToken` to every other POSTMAN request under Authorization

- Type: Bearer Token
- Token: `eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJlbWFpbCI6ImpzbzE5MTlAZ21haWwuY29tIiwiYWNjZXNzX.................0dNjX6X_yCjdt1PB4Hf4NE1wuvBWCXD5aD9GD72QuDk`

## Then you should be able to use the rest of the endpoints,

> Remember if you don't go through this process you will not be able to run any of the other endpoints

---

## IMPORTANT

You don't need the Production MONGO_URI to run the project, you should be able to run the API with the seeded data

## Tech Stack

![Mongoose](https://img.shields.io/badge/Mongoose-61DAFB?logo=Mongoose&logoColor=black)
![Node](https://img.shields.io/badge/Node-61DAFB?logo=Node&logoColor=black)
![TypeScript](https://img.shields.io/badge/TypeScript-3178C6?logo=TypeScript&logoColor=white)


CHANGELOG.md
#### 1.0.0 (2023-06-26)

##### Chores

* **server:**  updated readme with unit test and configuration (48ff3b7e)

##### New Features

*  add new husky script (e39f0bf6)
*  adding logs with changes (5f64feec)
*  adding logs with changes (fe8d4589)

##### Bug Fixes

*  added items (1cfcf45d)
*  add genereate-changelog as dev dependency (414898e2)
*  update CHANGELOG.md (6e4d6ee6)
*  remove github action (1de48deb)
*  update release yaml (b3769380)
*  update release yaml (06a35bde)
*  update release yaml (f8f201f5)

##### Other Changes

* bocacode/bocacode-intranet-api-typescript into 28-add-unit-tests-restaurants (01589821)
* true to disable and more fail-safes (f6d8ba01)
*  created and tested endpoints to post, get, update and disable, pending auth (efbead6e)
* **restaurants:**  added feature to add and edit restaurants (a5f8d69c)
* **user and logs:**  updated the updateUser routes and controller, added logs (a955680e)



src/scripts/seedData.ts
import { createRandomId } from '../utils/utils'

export const dummyRestaurants = [
  {
    uid: createRandomId(),
    address: '7024 Beracasa Way, Boca Raton, FL 33433',
    rating: '3',
    photo_url:
      'https://firebasestorage.googleapis.com/v0/b/bocacode-portal.appspot.com/o/restaurants%2Fwinn.jpeg?alt=media&token=cee866b1-b3ad-4881-8779-3511772cf751',
    menu_url: 'https://www.winndixie.com/shopping/departments/deli',
    created_by: 'data seed',
    name: 'Winn-Dixie',
  },
  {
    uid: createRandomId(),
    rating: '4',
    address: '7032 W Palmetto Park Rd, Boca Raton, FL 33433',
    created_by: 'data seed',
    name: 'Starbucks ',
    menu_url: 'starbucks.com/menu',
    photo_url:
      'https://firebasestorage.googleapis.com/v0/b/bocacode-portal.appspot.com/o/restaurants%2Fstarbucks%20.jpeg?alt=media&token=404140bf-3139-4ad4-bd24-24de58e34ddf',
  },
  {
    uid: createRandomId(),
    rating: '3',
    menu_url: 'https://www.chipotle.com/order/#menu',
    photo_url:
      'https://firebasestorage.googleapis.com/v0/b/bocacode-portal.appspot.com/o/restaurants%2Fchipotle.jpeg?alt=media&token=6e0caaf6-82fd-45c4-9ea2-60441d8ff001',
    address: '7028 W Palmetto Park Rd Ste 106, Boca Raton, FL 33433',
    created_by: 'data seed',
    name: 'Chipotle ',
  },
  {
    uid: createRandomId(),
    rating: '3',
    menu_url: 'https://www.publix.com/mc/order-ahead/order-subs-and-wraps',
    address: '7060 W Palmetto Park Rd, Boca Raton, FL 33433',
    created_by: 'data seed',
    name: 'Publix ',
    photo_url:
      'https://firebasestorage.googleapis.com/v0/b/bocacode-portal.appspot.com/o/restaurants%2Fpublix.png?alt=media&token=930bda94-69f6-4a61-9626-9a73eda90eaf',
  },
]

export const dummyHomework = [
  {
    uid: createRandomId(),
    created_by: 'data seed',
    enabled: true,
    week: 'Week 1',
    subtopic: '',
    topic: 'Objects',
    question: 'Create an object called student and add 3 properties with their values.',
  },
  {
    uid: createRandomId(),
    created_by: 'data seed',
    enabled: true,
    topic: 'Functions',
    question:
      'Write a function named assignGrade, that takes one argument, a number score. Return a grade for the score, either “A,” “B,” “C,” “D,” or “F.” Invoke that function for a few different scores and log the result to make sure it works.',
    week: 'Week 1',
    subtopic: 'Conditionals',
  },
  {
    uid: createRandomId(),
    created_by: 'data seed',
    enabled: true,
    subtopic: '',
    week: 'Week 1',
    question:
      'Create a function that takes an array of numbers and returns a new array with the square of each number.',
    topic: 'Functions',
  },
  {
    uid: createRandomId(),
    created_by: 'data seed',
    enabled: true,
    week: 'Week 1',
    question:
      'Create a function which returns the number of true values there are in an array. const testArray = [true, false, false, true, false, false, false, true, true, true, false].',
    subtopic: 'Arrays',
    topic: 'Functions',
  },
]

export const dummyTutorials = [
  {
    uid: createRandomId(),
    created_by: 'data seed',
    media_type: 'video',
    enabled: true,
    url: 'https://www.youtube.com/watch?v=UmSpfdxu3ro&ab_channel=DevDreamer',
    description: 'JavaScript data-types, explained by Dev Dreamer',
    title: 'Data-Types in JavaScript',
    topic: 'JavaScript',
  },
  {
    uid: createRandomId(),
    created_by: 'data seed',
    media_type: 'video',
    enabled: true,
    title: 'Error Handling in JavaScript',
    description: 'JavaScript error handling, explained by Bro Code',
    topic: 'JavaScript',
    url: 'https://www.youtube.com/watch?v=Z2l3cQ7bvBA&ab_channel=BroCode',
  },
  {
    uid: createRandomId(),
    created_by: 'data seed',
    media_type: 'video',
    enabled: true,
    topic: 'JavaScript',
    url: 'https://www.youtube.com/watch?v=T8tilXV03UI&ab_channel=DevDreamer',
    description: 'JavaScript function-expressions, explained by Dev Dreamer',
    title: 'Function-Expressions in JavaScript',
  },
  {
    uid: createRandomId(),
    created_by: 'data seed',
    media_type: 'video',
    enabled: true,
    url: 'https://www.youtube.com/watch?v=2-ojI5M9HME&ab_channel=BroCode',
    topic: 'JavaScript',
    title: 'Arrays in JavaScript',
    description: 'JavaScript arrays for beginners, explained by Bro Code',
  },
  {
    uid: createRandomId(),
    created_by: 'data seed',
    media_type: 'video',
    enabled: true,
    topic: 'JavaScript',
    title: 'Arrow-Functions in JavaScript',
    url: 'https://www.youtube.com/watch?v=Oy185MF8pnY&ab_channel=DevDreamer',
    description: 'JavaScript arrow functions, explained by Dev Dreamer',
  },
]
